### PR TITLE
Create registration middleware changed folders

### PR DESCRIPTION
**Pull request by valeriia-danylenko**
changed MVC pattern
added middlewares and errors
created registration

### CODE REVIEW

File: Demo2.1/app.js [hunk 1]
```diff
@@ -10,11 +15,14 @@ app.use(bodyParser.urlencoded({
 
 app.use(bodyParser.json())
 
+app.use("/register", registerRouter);
 app.use("/products", productsRouter);
-app.use("/order", ordersRouter);
+app.use("/order", validator.headers(loginUserDto), authMiddleware, ordersRouter);
```
Comment (iseldrewGD) [inline Demo2.1/app.js:20]: Please, use single quotes

File: Demo2.1/products/products_router.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const express = require('express');
+const productsController = require('./products_controller');
+const productsRouter = express.Router();
+const {searchParamsDto, idDto} = require('../common/other/dto.js');
+const validator = require('express-joi-validation').createValidator({});
+
+productsRouter.use("/search", validator.query(searchParamsDto), productsController.selectSearchedProducts);
+productsRouter.use("/:id", validator.params(idDto), productsController.selectProductDetails);
+productsRouter.use("/", productsController.selectAllProducts);
```
Comment (iseldrewGD) [inline Demo2.1/products/products_router.js:9]: Please, specify request type.

File: Demo2.1/common/other/dto.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+})
+
+module.exports = {
+    createUserDto,
+    loginUserDto,
+    productsArrDto,
+    searchParamsDto,
+    idDto
+};
+
```
Comment (iseldrewGD) [inline Demo2.1/common/other/dto.js:42]: I think would be better to split dtos between separated files.

File: Demo2.1/common/other/errors.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        this.data = data;
+    }
+};
+
+module.exports = {
+    Forbidden,
+    NotFound,
+    CustomError,
+    NotAcceptable
+}
```
Comment (iseldrewGD) [inline Demo2.1/common/other/errors.js:39]: I think would be better to split exceptions between separated files.

File: Demo2.1/orders/orders_model.js [hunk 5]
```diff
@@ -19,7 +19,7 @@ class OrdersModel extends DatabasePool {
 
     validateProductsArr(products) {
         const invalidProducts = products.filter(product => {
-            return typeof product.id !== 'number' || typeof product.count !== 'number' || product.id <= 0 || product.count <=0
+            return typeof product.id !== 'number' || typeof product.count !== 'number' || product.id <= 0 || product.count <= 0
         })
         return invalidProducts
     }
```
Comment (iseldrewGD) [inline Demo2.1/orders/orders_model.js:25]: I guess, we don't need this method anymore.

File: Demo2.1/register/register_model.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+class RegisterModel extends DatabasePool {
+
+    async checkNew(user) {
+        let { phone } = user
+        const foundUser = await ordersModel.searchUser(phone);
+        if (foundUser.length > 0) {
+            return false
+        }
+        return true
```
Comment (iseldrewGD) [inline Demo2.1/register/register_model.js:11]:
```suggestion
        return !foundUser.length
```

**Review by iseldrewGD — COMMENTED** [review]
Add newline to the end of files.

File: Demo2.1/app.js [hunk 7]
```diff
@@ -10,11 +15,14 @@ app.use(bodyParser.urlencoded({
...
 
 app.use(function (req, res, next) {
     res.status(404).send("Not Found")
 });
+
+app.use(errorHandler);
  
 app.listen(3000);
```
Comment (kolesnicknick) [inline Demo2.1/app.js:28]: we should change it on environment variable || 3000

File: Demo2.1/app.js [hunk 8]
```diff
@@ -10,11 +15,14 @@ app.use(bodyParser.urlencoded({
...
 app.use(bodyParser.json())
 
+app.use("/register", registerRouter);
 app.use("/products", productsRouter);
-app.use("/order", ordersRouter);
+app.use("/order", validator.headers(loginUserDto), authMiddleware, ordersRouter);
 
 app.use(function (req, res, next) {
     res.status(404).send("Not Found")
 });
```
Comment (kolesnicknick) [inline Demo2.1/app.js:24]: Here we should return our newly created error =)

File: Demo2.1/common/middleware/error_middleware.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+            error: err.error || 'Server error',
+            data: err.data
+        });
+    } else {
+        res.status(err.statusCode).json({
+            success: false,
+            error: err.error || 'Server error'
+        });
+    };
+};
```
Comment (kolesnicknick) [inline Demo2.1/common/middleware/error_middleware.js:15]: we can remove else, if data is not present, we will not see it in response anyway as it is undefined

File: Demo2.1/orders/orders_model.js (hunk 5 above)
Comment (kolesnicknick) [inline Demo2.1/orders/orders_model.js:25]: can we move this validations to Joi?

File: Demo2.1/register/register_model.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+const DatabasePool = require('../common/other/pool_model');
+
+class RegisterModel extends DatabasePool {
+
+    async checkNew(user) {
```
Comment (kolesnicknick) [inline Demo2.1/register/register_model.js:5]: change name of function to verbose definition:
checkNew is not obvious name
checkUserExistByPhone is much better

File: Demo2.1/register/register_model.js (hunk 6 above)
Comment (kolesnicknick) [inline Demo2.1/register/register_model.js:11]: if user exists(length > 0) we return false? could you describe a bit more about the whole logic?

**Review by kolesnicknick — COMMENTED** [review]
great job!

**Pull request merged**